Add Marketplace documentation

dennisameling opened the pull request:
Belongs to https://github.com/mautic/mautic/pull/10518

<a href="https://gitpod.io/#https://github.com/mautic/developer-documentation-new/pull/10"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

RCheesley commented on docs/marketplace/listing.rst:
```diff
@@ -0,0 +1,71 @@
+- ``type`` must have value ``mautic-plugin``. The Marketplace is filtering PHP packages by this tag. It is required to show up in the Marketplace.
+- ``extra.install-directory-name`` specifies the directory name for the bundle. Correct directory name is important for PSR4 autoloading. It must be the same as in the namespace in your PHP classes.
+- ``require.php`` be sure to specify the PHP version range that you test the plugin against. Ideally it should be the same as the Mautic's `supported PHP versions <https://www.mautic.org/download/requirements>`_. Do not allow users to install the plugin on versions that are not supported.
+- ``require.ext-*`` if your plugin require some PHP extension, please list them in the require section too. No one knows on what servers your plugin will run on.
```
```suggestion
- ``require.ext-*`` if your Plugin require some PHP extension, please list them in the require section too. As servers can have a wide variety of extensions enabled, this ensures that the correct extensions are available.
```

RCheesley commented on docs/marketplace/listing.rst:
```diff
@@ -0,0 +1,71 @@
+- ``extra.install-directory-name`` specifies the directory name for the bundle. Correct directory name is important for PSR4 autoloading. It must be the same as in the namespace in your PHP classes.
+- ``require.php`` be sure to specify the PHP version range that you test the plugin against. Ideally it should be the same as the Mautic's `supported PHP versions <https://www.mautic.org/download/requirements>`_. Do not allow users to install the plugin on versions that are not supported.
+- ``require.ext-*`` if your plugin require some PHP extension, please list them in the require section too. No one knows on what servers your plugin will run on.
+- ``require.mautic/core-lib`` it's important to specify which Mautic versions your plugin supports. Include only the versions you or your community have tested and confirmed the plugin works with.
```
```suggestion
- ``require.mautic/core-lib`` it's important to specify which Mautic versions your Plugin supports. Include only the versions you or your community have tested and confirmed the Plugin works with.
```

RCheesley commented on docs/marketplace/listing.rst:
```diff
@@ -0,0 +1,71 @@
+- ``require.ext-*`` if your plugin require some PHP extension, please list them in the require section too. No one knows on what servers your plugin will run on.
+- ``require.mautic/core-lib`` it's important to specify which Mautic versions your plugin supports. Include only the versions you or your community have tested and confirmed the plugin works with.
+
+Step 2: check best practices
```
```suggestion
Step 2: verify best practices
```

RCheesley commented on docs/marketplace/listing.rst:
```diff
@@ -0,0 +1,71 @@
+Step 2: check best practices
+----------------------------
+
+There's a few things we highly recommend when creating plugins for Mautic. Please refer to the :doc:`./best_practices` page for more details.
```
```suggestion
There's a few things which are highly recommend when creating Plugins for Mautic. Please refer to the :doc:`./best_practices` page for more details.
```

RCheesley commented on docs/marketplace/listing.rst:
```diff
@@ -0,0 +1,71 @@
+Step 3: publish to Composer
+---------------------------
+
+When the composer.json is ready, follow the `Publishing Packages <https://packagist.org>`_ section directly in Packagist.
```
Please use the external linking extension here!

RCheesley commented on docs/marketplace/listing.rst:
```diff
@@ -0,0 +1,71 @@
+Step 4: apply for the allowlist
+-------------------------------
+
+While the Mautic Marketplace is still in beta, you'll have to apply for our allowlist in order to show up in the Marketplace. We expect to move to stable with Mautic 4.2 in February 2022, after which the allowlist likely won't be needed anymore. Please refer to the :doc:`./allowlist_what_and_why` page for more details.
```
```suggestion
While the Mautic Marketplace is still in beta, you'll have to apply for our allowlist in order to show up in the Marketplace. It is expected that the Marketplace will move from beta to stable release with Mautic 4.2 in February 2022, after which the allowlist may be deprecated in favour of a blocklist. Please refer to the :doc:`./allowlist_what_and_why` page for more details.
```

RCheesley commented on docs/marketplace/listing.rst:
```diff
@@ -0,0 +1,71 @@
+
+While the Mautic Marketplace is still in beta, you'll have to apply for our allowlist in order to show up in the Marketplace. We expect to move to stable with Mautic 4.2 in February 2022, after which the allowlist likely won't be needed anymore. Please refer to the :doc:`./allowlist_what_and_why` page for more details.
+
+Updating existing plugins for usage with the Marketplace
```
```suggestion
Updating existing Plugins for usage with the Marketplace
```

RCheesley commented on docs/marketplace/listing.rst:
```diff
@@ -0,0 +1,71 @@
+
+If you required ``mautic/composer-plugin`` in your plugin's dependencies in the past, please remove it. Support for Mautic plugins is now built into Composer, so you only have to set the type to ``mautic-plugin`` and Composer will automatically install your plugin into the ``plugins`` folder.
+
+Next to that, if you built your plugin for Mautic 3 originally, please check our `UPGRADE-4.0.md guide <https://github.com/mautic/mautic/blob/4.x/UPGRADE-4.0.md>`_ for the breaking changes we made in Mautic 4.
```
```suggestion
Next to that, if you built your plugin for Mautic 3 originally, please check the `UPGRADE-4.0.md guide <https://github.com/mautic/mautic/blob/4.x/UPGRADE-4.0.md>`_ for the breaking changes in Mautic 4.
```

RCheesley commented:
@dennisameling should also mention that Vale seems to be passing despite an awful lot of warnings - is it configured correctly?

dennisameling commented:
@RCheesley 
- Added a build + checklinks step to CI - this ensures that external links can actually be reached (404/500/etc. error codes will fail the job)
- Added Vale + several plugins like GitLens to Gitpod, so that contributors see the warnings/suggestions right within their editor while working on docs
- Moved external links to their own files as requested
- Fixed several warnings & suggestions

There's two small TODOs left in this PR - adding screenshots and linking to the plugin creation docs (which I'll create in a separate PR)

RCheesley commented on docs/marketplace/allowlist_what_and_why.rst:
```diff
@@ -0,0 +1,27 @@
+============================================
+Allowlist: what's it and why is it needed?
```
```suggestion
Allowlist: what is it and why is it needed?
```
Guessing this was a Vale correction but actually makes better grammatical sense to not abbreviate in that way.

RCheesley reviewed (changes requested):
One small fix to grammar otherwise good to go. Let's not forget the TODOs!

Also as a side note, when I spun up Gitpod it asked me to install Python (the extension maybe?) and it says Vale is not available. Maybe something is not quite right there?
